cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll

```diff
@@ -101,12 +106,18 @@ class ValueNumber extends TValueNumber {
  * The use of `p.x` on line 3 is linked to the definition of `p` on line 1 as well, but is not
  * congruent to that definition because `p.x` accesses only a subset of the memory defined by `p`.
  */
-private class CongruentCopyInstruction extends CopyInstruction {
-  CongruentCopyInstruction() {
+class CongruentCopyInstructionExact extends CopyInstruction {
+  CongruentCopyInstructionExact() {
     this.getSourceValueOperand().getDefinitionOverlap() instanceof MustExactlyOverlap
   }
 }
 
+class CongruentCopyInstructionTotal extends CopyInstruction {
+  CongruentCopyInstructionTotal() {
+    this.getSourceValueOperand().getDefinitionOverlap() instanceof MustTotallyOverlap
```

Comment:
What happens if this restriction is relaxed to include `MayPartiallyOverlap`?

Reply:
There _should_ be no operands for which `getDefinitionOverlap()` returns `MayPartiallyOverlap`. The main reason `Chi` instructions exist is to ensure that a definition always totally overlaps its uses.